### PR TITLE
Implemet possibility to reroute to error

### PR DESCRIPTION
**Pull request by caalador**
Now there is a possibility to reroute to an
error target from beforeNavigation

---
This change is [Reviewable](https://reviewable.io/reviews/vaadin/flow/2649)

### CODE REVIEW

**Comment by denis-anisimov** [comment]
Reviewed 6 of 6 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[flow-documentation/routing/tutorial-routing-exception-handling.asciidoc, line 78 at r1](https://reviewable.io:443/reviews/vaadin/flow/2649#-Kw9kIbHB-bmEuMi5igQ:-Kw9kIbHB-bmEuMi5igR:buj59py) ([raw file](https://github.com/vaadin/flow/blob/6c02d5a9ff2c33aba2aae8e46f3c3b100b2baa15/flow-documentation/routing/tutorial-routing-exception-handling.asciidoc#L78)):*
> The rerouting is done by using one of the overloads for `BeforeNavigationEvent::rerouteToError` with only the
> exception class to target or with an added custom error message.

It would be nice to see some example here

---

*[flow-server/src/main/java/com/vaadin/router/event/BeforeNavigationEvent.java, line 269 at r1](https://reviewable.io:443/reviews/vaadin/flow/2649#-Kw9lFq6_kaRs_-9Opnv:-Kw9lFq6_kaRs_-9Opnw:bj7fmph) ([raw file](https://github.com/vaadin/flow/blob/6c02d5a9ff2c33aba2aae8e46f3c3b100b2baa15/flow-server/src/main/java/com/vaadin/router/event/BeforeNavigationEvent.java#L269)):*
>  Reroute to error target for given exception without custom message.

Same as below: default CTOR for exception class .

Also `@see` since it's a shorthand.

---

*[flow-server/src/main/java/com/vaadin/router/event/BeforeNavigationEvent.java, line 279 at r1](https://reviewable.io:443/reviews/vaadin/flow/2649#-Kw9l8NnqLAIvSzSCLHa:-Kw9l8NnqLAIvSzSCLHb:b-vva567) ([raw file](https://github.com/vaadin/flow/blob/6c02d5a9ff2c33aba2aae8e46f3c3b100b2baa15/flow-server/src/main/java/com/vaadin/router/event/BeforeNavigationEvent.java#L279)):*
> Reroute to error target for given exception with given custom message.

It should mention that the exception class should have default CTOR.

`@see`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/2649)*

**Comment by caalador** [comment]
Review status: 4 of 7 files reviewed at latest revision, 3 unresolved discussions.

---

*[flow-documentation/routing/tutorial-routing-exception-handling.asciidoc, line 78 at r1](https://reviewable.io:443/reviews/vaadin/flow/2649#-Kw9kIbHB-bmEuMi5igQ:-Kw9riV_Pp5lhcFuOUW9:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/6c02d5a9ff2c33aba2aae8e46f3c3b100b2baa15/flow-documentation/routing/tutorial-routing-exception-handling.asciidoc#L78)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> The rerouting is done by using one of the overloads for `BeforeNavigationEvent::rerouteToError` with only the
> exception class to target or with an added custom error message.

It would be nice to see some example here
</blockquote></details>

Done.

---

*[flow-server/src/main/java/com/vaadin/router/event/BeforeNavigationEvent.java, line 269 at r1](https://reviewable.io:443/reviews/vaadin/flow/2649#-Kw9lFq6_kaRs_-9Opnv:-Kw9r2aRGu8aaL815kTb:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/6c02d5a9ff2c33aba2aae8e46f3c3b100b2baa15/flow-server/src/main/java/com/vaadin/router/event/BeforeNavigationEvent.java#L269)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

>  Reroute to error target for given exception without custom message.

Same as below: default CTOR for exception class .

Also `@see` since it's a shorthand.
</blockquote></details>

Done.

---

*[flow-server/src/main/java/com/vaadin/router/event/BeforeNavigationEvent.java, line 279 at r1](https://reviewable.io:443/reviews/vaadin/flow/2649#-Kw9l8NnqLAIvSzSCLHa:-Kw9r2OReF_0nuZMhoj3:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/6c02d5a9ff2c33aba2aae8e46f3c3b100b2baa15/flow-server/src/main/java/com/vaadin/router/event/BeforeNavigationEvent.java#L279)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> Reroute to error target for given exception with given custom message.

It should mention that the exception class should have default CTOR.

`@see`
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/2649)*

**Comment by denis-anisimov** [comment]
Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/2649)*

**Comment by denis-anisimov** [comment]
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[flow-server/src/main/java/com/vaadin/router/event/BeforeNavigationEvent.java, line 51 at r3](https://reviewable.io:443/reviews/vaadin/flow/2649#-Kw9yYbb7IaxK_SXNH6x:-Kw9yYbb7IaxK_SXNH6y:b-1ychm3) ([raw file](https://github.com/vaadin/flow/blob/7fe0053768dfdaa4f34a4a634246f8523eac5c55/flow-server/src/main/java/com/vaadin/router/event/BeforeNavigationEvent.java#L51)):*
> private transient ErrorParameter<?> errorParameter;

So after the serialization it will be `null` . Is it expected ?
`ErrorParameter` should be serializable.
Ot just ignore it for now at all since there are many places where non-serializable classes are used.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/2649)*

**Comment by caalador** [comment]
Review status: 5 of 7 files reviewed at latest revision, 7 unresolved discussions.

---

*[flow-server/src/main/java/com/vaadin/router/event/BeforeNavigationEvent.java, line 51 at r3](https://reviewable.io:443/reviews/vaadin/flow/2649#-Kw9yYbb7IaxK_SXNH6x:-Kw9zh0jSxUt9jL3B8NN:by0mkp4) ([raw file](https://github.com/vaadin/flow/blob/7fe0053768dfdaa4f34a4a634246f8523eac5c55/flow-server/src/main/java/com/vaadin/router/event/BeforeNavigationEvent.java#L51)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> private transient ErrorParameter<?> errorParameter;

So after the serialization it will be `null` . Is it expected ?
`ErrorParameter` should be serializable.
Ot just ignore it for now at all since there are many places where non-serializable classes are used.
</blockquote></details>

Made it serializable.
Basically we should never have it so that it would be serialized.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/2649)*

**Comment by denis-anisimov** [comment]
Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/2649)*

**Comment by denis-anisimov** [comment]
Reviewed 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/2649)*

**Comment by torok-peter** [comment]
Reviewed 1 of 6 files at r1, 2 of 3 files at r2, 1 of 2 files at r4, 1 of 2 files at r5, 3 of 3 files at r6.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[flow-server/src/main/java/com/vaadin/router/Router.java, line 165 at r6](https://reviewable.io:443/reviews/vaadin/flow/2649#-KwAla-GDq1PacJNvYy8:-KwAla-H_zycTI69tnH8:b-pujj8v) ([raw file](https://github.com/vaadin/flow/blob/071f8b372b167347907d452e5d80e4e35854de7b/flow-server/src/main/java/com/vaadin/router/Router.java#L165)):*
> ```Java
>             throw new RuntimeException(errorParameter.getCustomMessage(),
>                     errorParameter.getException());
>         }
> ```

Could be simplified as
```
Class<? extends Component> navigationTarget = getRegistry()
                .getErrorNavigationTarget(errorParameter.getException())
                .orElseThrow(() -> new RuntimeException(errorParameter.getCustomMessage(),
                    errorParameter.getException()));

ErrorStateRenderer handler = ...
```

---

*[flow-server/src/main/java/com/vaadin/router/event/BeforeNavigationEvent.java, line 337 at r6](https://reviewable.io:443/reviews/vaadin/flow/2649#-KwAm_-kKVxhLUFStQ2S:-KwAm_-ltsAsHRnBFf7J:b6zj7gk) ([raw file](https://github.com/vaadin/flow/blob/071f8b372b167347907d452e5d80e4e35854de7b/flow-server/src/main/java/com/vaadin/router/event/BeforeNavigationEvent.java#L337)):*
> ```Java
>             throw new RuntimeException(customMessage,
>                     errorParameter.getException());
>         }
> ```

Could be simplified using `orElseThrow` as above

---

*[flow-server/src/test/java/com/vaadin/router/RouterTest.java, line 1401 at r6](https://reviewable.io:443/reviews/vaadin/flow/2649#-KwAn5gn2JANkCuBLUzA:-KwAn5gn2JANkCuBLUzB:b-oil5hd) ([raw file](https://github.com/vaadin/flow/blob/071f8b372b167347907d452e5d80e4e35854de7b/flow-server/src/test/java/com/vaadin/router/RouterTest.java#L1401)):*
> ```Java
> 
>     @Test
>     public void reroute_to_error_with_custom_message_message_is_used()
> ```

"message" seems to be duplicated?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/2649)*

**Comment by caalador** [comment]
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[flow-server/src/main/java/com/vaadin/router/Router.java, line 165 at r6](https://reviewable.io:443/reviews/vaadin/flow/2649#-KwAla-GDq1PacJNvYy8:-KwDs9v8pKeUnw4Pc3Eu:bct6jw0) ([raw file](https://github.com/vaadin/flow/blob/071f8b372b167347907d452e5d80e4e35854de7b/flow-server/src/main/java/com/vaadin/router/Router.java#L165)):*
<details><summary><i>Previously, torok-peter (Péter Török) wrote…</i></summary><blockquote>

Could be simplified as
```
Class<? extends Component> navigationTarget = getRegistry()
                .getErrorNavigationTarget(errorParameter.getException())
                .orElseThrow(() -> new RuntimeException(errorParameter.getCustomMessage(),
                    errorParameter.getException()));

ErrorStateRenderer handler = ...
```
</blockquote></details>

You say simplified, I say not as readable.

---

*[flow-server/src/main/java/com/vaadin/router/event/BeforeNavigationEvent.java, line 337 at r6](https://reviewable.io:443/reviews/vaadin/flow/2649#-KwAm_-kKVxhLUFStQ2S:-KwDtTYMAOULz8OjOsjX:b-uysgcd) ([raw file](https://github.com/vaadin/flow/blob/071f8b372b167347907d452e5d80e4e35854de7b/flow-server/src/main/java/com/vaadin/router/event/BeforeNavigationEvent.java#L337)):*
<details><summary><i>Previously, torok-peter (Péter Török) wrote…</i></summary><blockquote>

Could be simplified using `orElseThrow` as above
</blockquote></details>

One liner makes it harder to comprehend.

---

*[flow-server/src/test/java/com/vaadin/router/RouterTest.java, line 1401 at r6](https://reviewable.io:443/reviews/vaadin/flow/2649#-KwAn5gn2JANkCuBLUzA:-KwDsTPUNbme88_7XCIu:b-qyioii) ([raw file](https://github.com/vaadin/flow/blob/071f8b372b167347907d452e5d80e4e35854de7b/flow-server/src/test/java/com/vaadin/router/RouterTest.java#L1401)):*
<details><summary><i>Previously, torok-peter (Péter Török) wrote…</i></summary><blockquote>

"message" seems to be duplicated?
</blockquote></details>

it reads do:`Reroute to error with custom message` result `message is used`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/2649)*

**Comment by torok-peter** [comment]
Reviewed 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/2649)*
